### ISSUE TITLE
Exposing Calendar arithmetic again

### ISSUE THREAD

**Issue by Ekleog** [unicode-org/icu4x#3082]
Hey!

I've learned of icu4x only very recently, but it looks awesome! And I learned it at the exact right time, when I'm trying to add features for formatting human-readable times in my (still WIP) chrono competitor crate.

I noticed that the date arithmetic APIs [have been hidden on purpose for 1.0](https://github.com/unicode-org/icu4x/issues/1675).

Do you know what re-exposing them is blocked on?

Anyway, thank you for icu4x, even without calendar arithmetic it's already looking great! :)

**Comment by sffc** [comment]
It was set aside because we didn't have the bandwidth to make it a fully polished feature in time for 1.0; if you are interested in this subject area, there are some low-hanging tasks that you could contribute if you like!

**Comment by Ekleog** [comment]
Makes sense! I'm interested indeed, and would love to contribute if there are things I could :)

(Disclaimer: I'll still focus on finishing up the rest of my crate first, as I also need to implement timezones still too, but I'd love to know what's missing so I could work on it as soon as that's done :))

**Comment by Ekleog** [comment]
FWIW, the thoughts I already had while thinking about calendar arithmetic, before learning that it's in the long-term scope of icu4x, with comments relating to what I read in icu4x:
- Addition and subtraction are not commutative. Adding one day then one month can result in a different value than one month then one day (eg. `Apr 30 + 1d + 1mo = Jun 1`, but `Apr 30 + 1mo + 1d = May 31`)
- So it makes sense to introduce addition of a fix number of days / weeks / months / years (/ eras? not entirely sure about that), but doing so non-commutatively. This to me, for best user experience, hints at insertion of `Days(isize)`, `Weeks(isize)` etc. types. They can be added/subtracted between themselves and with a `Date`, but not between themselves; this way the commutativity properties are kept.
- In particular, I'm not sure the current `DateDuration` type makes sense. This particular behavior is visible in your current [`test_offset_handles_out_of_bound_month_offset`](https://github.com/unicode-org/icu4x/blob/main/components/calendar/src/iso.rs#L658): it arbitrarily decided that addition should be first month then day, but I'm not sure it'd make more sense than the opposite. Especially when considering the current "month addition overflows into day" behavior.
- Even with separate types for days and months, there is the question of how to round when getting out of the range of the month. Right now, the test just carries over the missing days to the next month (`Jan 31 + 1mo = Mar 3`). Alternatives would be staying in the month (so returning `Feb 28`), or considering out-of-month days as "fractions" and rounding to `Mar 1`.
TBH I'm not sure I can think of which one would be best to me, but I must say that I expected the result to be either `Feb 28` or `Mar 1`, and seeing the test return `Mar 3` surprised me: it feels as though addition was handling by adding a number of days, but the length of a month is not a fixed number of days and that's precisely why we go through all these troubles. That said I also do understand that some use cases would need this handling, so… I don't know? Maybe `Date + Month` should not exist, but only specific `Date.truncating_add(Month)`, `Date.rounding_add(Month)` and `Date.overflowing_add(Month)`?

With this in mind, the things that are missing that I personally see as being needed are:
- Split `DateDuration` into `Days`, `Months`, etc. types (if you agree with me that it's a good idea)
- Decide on what month overflow behavior should be. Unfortunately, it looks like [ICU documentation does not address this question](https://unicode-org.github.io/icu/userguide/datetime/calendar/#field-arithmetic) (or it does at a place I didn't know I should check). I'm personally in favor of not having `+` and having different methods for each kind of rounding, but please let me know your thoughts :)

I'm willing to get to these whenever you tell me what your thoughts are on the matter, and please let me know if there are other things you think should be done on the calendar API! :)

(BTW, I saw on the ICU website, while looking for whether there is a defined month overflow behavior there, that ICU is also supposed to be dealing with timezones. I'm currently writing code supporting this for `kine`, but would love to work with you to try and merge that with `icu4x` if it has not just been decided as being out-of-scope)

**Comment by Manishearth** [comment]
I was under the impression we had a proper issue for this, but it seems like we don't. Filing https://github.com/unicode-org/icu4x/issues/3147



(Closing this issue since it's a question)

> * In particular, I'm not sure the current `DateDuration` type makes sense. This particular behavior is visible in your current [`test_offset_handles_out_of_bound_month_offset`](https://github.com/unicode-org/icu4x/blob/main/components/calendar/src/iso.rs#L658): it arbitrarily decided that addition should be first month then day, but I'm not sure it'd make more sense than the opposite. Especially when considering the current "month addition overflows into day" behavior.

We're matching Temporal behavior here.


> * Even with separate types for days and months, there is the question of how to round when getting out of the range of the month. Right now, the test just carries over the missing days to the next month (`Jan 31 + 1mo = Mar 3`). Alternatives would be staying in the month (so returning `Feb 28`), or considering out-of-month days as "fractions" and rounding to `Mar 1`.

Yeah I think the spec clamps/rejects instead ([by calling RegulateISODate](https://tc39.es/proposal-temporal/#sec-temporal-regulateisodate)). We should be doing that.
